src/dune/sub_dirs.ml

```diff
@@ -112,34 +130,106 @@ let eval (t : _ Status.Map.t) ~dirs =
                 dir
             ]))
 
-let decode =
+type subdir_stanzas = (Loc.t * Predicate_lang.Glob.t) option Status.Map.t
+
+module Dir_map = struct
+  type per_dir =
+    { sexps : Dune_lang.Ast.t list
+    ; subdir_status : subdir_stanzas
+    }
+
+  type t =
+    { data : per_dir
+    ; nodes : t String.Map.t
+    }
+
+  let empty_per_dir =
+    { sexps = []; subdir_status = Status.Map.init (fun _ -> None) }
+
+  let empty = { data = empty_per_dir; nodes = String.Map.empty }
+
+  let root t = t.data
+
+  let descend t (p : string) = String.Map.find t.nodes p
+
+  let sub_dirs t = String.Map.keys t.nodes
+
+  let rec make_at_path path data =
+    match path with
+    | [] -> data
+    | x :: xs ->
+      let nodes = String.Map.singleton x (make_at_path xs data) in
+      { empty with nodes }
+
+  let singleton data = { empty with data }
+
+  let merge_data d1 d2 =
+    { sexps = d1.sexps @ d2.sexps
+    ; subdir_status =
+        Status.Map.merge d1.subdir_status d2.subdir_status ~f:(fun l r ->
+            match (l, r) with
+            | acc, None
+            | None, acc ->
+              acc
+            | Some (loc, _), Some (loc2, _) ->
+              User_error.raise ~loc
+                [ Pp.text "This stanza stanza was already specified at:"
+                ; Pp.verbatim (Loc.to_file_colon_line loc2)
+                ])
+    }
+
+  let rec merge t1 t2 : t =
+    let data = merge_data t1.data t2.data in
+    let nodes =
+      String.Map.union t1.nodes t2.nodes ~f:(fun _ l r -> Some (merge l r))
+    in
+    { data; nodes }
+
+  let merge_all = List.fold_left ~f:merge ~init:empty
```

Comment:
This seems quite suboptimal. However, I have a feeling that we are unlikely to ever merge anything large in real projects.

@aalekseyev if you'd like to micro-optimize, this is a good candidate :)

Reply:
To me it seems fine to leave it as-is.